lsp/src/uri0.ml

```diff
@@ -1,31 +1,143 @@
+(* This module is based on the [vscode-uri] implementation:
+   https://github.com/microsoft/vscode-uri/blob/main/src/uri.ts. It only
+   supports scheme, authority and path. Query, port and fragment are not
+   implemented *)
+
 open Import
 
 module Private = struct
   let win32 = ref Sys.win32
 end
 
-type t = Uri_lexer.t =
-  { scheme : string option
+type t =
+  { scheme : string
   ; authority : string
   ; path : string
   }
 
-let t_of_yojson json = Json.Conv.string_of_yojson json |> Uri_lexer.of_string
+let backslash_to_slash =
+  String.map ~f:(function
+    | '\\' -> '/'
+    | c -> c)
+
+let slash_to_backslash =
+  String.map ~f:(function
+    | '/' -> '\\'
+    | c -> c)
+
+let add_prefix_if_not_exists s ~prefix =
+  if String.is_prefix s ~prefix then s else prefix ^ s
+
+let of_path path =
+  let path = if !Private.win32 then backslash_to_slash path else path in
+  let path, authority =
+    let len = String.length path in
+    if len = 0 then ("/", "")
+    else if String.is_prefix path ~prefix:"//" then
+      let offset = 2 in
+      let idx = String.index_from_opt path offset '/' in
+      match idx with
+      | None -> ("/", String.sub path ~pos:offset ~len:(len - offset))
+      | Some i ->
+        let authority = String.sub path ~pos:offset ~len:(i - offset) in
+        let path =
+          let path = String.sub path ~pos:i ~len:(len - i) in
+          if String.is_empty path then "/" else path
+        in
+        (path, authority)
+    else (path, "")
+  in
+  let path = add_prefix_if_not_exists path ~prefix:"/" in
+  { scheme = "file"; authority; path }
+
+let to_path { path; authority; scheme } =
+  let path =
+    let len = String.length path in
+    if len = 0 then "/"
+    else
+      let buff = Buffer.create 64 in
+      (if (not (String.is_empty authority)) && len > 1 && scheme = "file" then (
+       Buffer.add_string buff "//";
+       Buffer.add_string buff authority;
+       Buffer.add_string buff path)
+      else if len < 3 then Buffer.add_string buff path
+      else
+        let c0 = path.[0] in
+        let c1 = path.[1] in
+        let c2 = path.[2] in
+        if
+          c0 = '/'
+          && ((c1 >= 'A' && c1 <= 'Z') || (c1 >= 'a' && c1 <= 'z'))
+          && c2 = ':'
+        then (
+          Buffer.add_char buff (Char.lowercase_ascii c1);
+          Buffer.add_string buff
+            (String.sub path ~pos:2 ~len:(String.length path - 2)))
+        else Buffer.add_string buff path);
+      Buffer.contents buff
+  in
+  if !Private.win32 then slash_to_backslash path else path
+
+let of_string s =
+  let Uri_lexer.{ scheme; authority; path } = Uri_lexer.of_string s in
+  let scheme = scheme |> Option.value ~default:"file" in
```

Comment:
This looks much better with ocamllex now. What about pushing this `scheme`, `authority`, and `path` post processing there as well?